# Pandas II

Now that we are familiar with loading data into a dataframe and looking at it, let's see how it can allow us to methodically analyize our data. 

We will cover:

Topic | Method
-|-
[picking out a column or columns by name](#Selecting-a-column) | `['col_name']`
[picking out a row or rows by index](#Selecting-a-row) | `[0:100] # this is rows 0 to 99`
[indexing (slicing or selecting) data by label](#Selecting-by-label) | `.loc[]`
[indexing (slicing or selecting) data by position](#Selecting-by-position) | `.iloc[]`


We will use the Nobel prize data that's accessible through the academy's api. We will begin by loading the data into a dataframe and viewing the metadata and the first few rows.


## *Resources*
[Indexing in pandas documentation](http://pandas.pydata.org/pandas-docs/stable/indexing.html)

In [1]:
%load_ext watermark
%watermark -a 'Alexander C Booth' -mv -p pandas

Alexander C Booth 

CPython 3.5.1
IPython 4.2.0

pandas 0.18.1

compiler   : GCC 4.2.1 Compatible Apple LLVM 4.2 (clang-425.0.28)
system     : Darwin
release    : 16.0.0
machine    : x86_64
processor  : i386
CPU cores  : 4
interpreter: 64bit


In [2]:
import pandas as pd

In [3]:
df = pd.read_csv('../../data/nobel_data.csv')

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 900 entries, 0 to 899
Data columns (total 8 columns):
firstname     900 non-null object
id            900 non-null int64
motivation    812 non-null object
share         900 non-null int64
surname       870 non-null object
year          900 non-null int64
gender        900 non-null object
prize         900 non-null object
dtypes: int64(3), object(5)
memory usage: 56.3+ KB


In [5]:
df.head()

firstname   id                                         motivation  share  \
0    Takaaki  919  "for the discovery of neutrino oscillations, w...      2   
1  Arthur B.  920  "for the discovery of neutrino oscillations, w...      2   
2      Isamu  906  "for the invention of efficient blue light-emi...      3   
3    Hiroshi  907  "for the invention of efficient blue light-emi...      3   
4      Shuji  908  "for the invention of efficient blue light-emi...      3   

    surname  year gender    prize  
0    Kajita  2015   male  physics  
1  McDonald  2015   male  physics  
2   Akasaki  2014   male  physics  
3     Amano  2014   male  physics  
4  Nakamura  2014   male  physics

From the info and the rows we viewed, we can see that there are 900 rows and 8 columns. For example, there is a column called 'prize' that indicates the type of prize and a gender column. We will uses these below.

## Indexing
In pandas, it is easy to subset our dataframe by indexing the rows and columns. We can say, "pandas give me the part of the dataframe that matches this row criteria and column criteria". We will see how below.

### Selecting a column
We can subset our dataframe to a particular column by calling `df['col_name']`. We can subset by multiple columns by putthing them into a list `df[['col1','col2','col3']]`.

In the example below, we will first subset just the prize column. Next we will then subset by both prize and gender.

***Note that we will also put the `.head()` method at the end of it so that it only prints out a few rows***

In [6]:
df['prize'].head()

0    physics
1    physics
2    physics
3    physics
4    physics
Name: prize, dtype: object

In [7]:
df[['prize', 'gender']].head() # Notice the double square brackets here since multiple columns require a list

prize gender
0  physics   male
1  physics   male
2  physics   male
3  physics   male
4  physics   male

### Selecting a row
We can similarly subset by a particular row. The rows are named by the index, which is usually a list of numbers or dates, or sometimes even labeled by names. We can access the index of a dataframe any time by calling the `.index` attribute. We will do this first below to confirm the index is a range of numbers and then use these numbers to pick out a particular row or rows.

In [8]:
df.index

RangeIndex(start=0, stop=900, step=1)

We see above that the index is a python range starting at 0 and stopping at 900, by increments of 1. We can think of this index being the list `[0,1,2..., ...898, 899]`. We must pass a list of numbers in the same way we would to index a list to get a row or rows.

*Note: the index doesn't have to be a numerical range -- it could just as easily be any labels you want. And it doesn't stop there, you can even have multiple indexes for one dataframe. More on this point in future lectures.*

In [9]:
df[0:1] # The very first row

firstname   id                                         motivation  share  \
0   Takaaki  919  "for the discovery of neutrino oscillations, w...      2   

  surname  year gender    prize  
0  Kajita  2015   male  physics

In [10]:
df[15:20] # Row 15-19

firstname   id                                         motivation  share  \
15  Willard S.  839  "for the invention of an imaging semiconductor...      4   
16   George E.  840  "for the invention of an imaging semiconductor...      4   
17    Yoichiro  826  "for the discovery of the mechanism of spontan...      2   
18      Makoto  827  "for the discovery of the origin of the broken...      4   
19   Toshihide  828  "for the discovery of the origin of the broken...      4   

      surname  year gender    prize  
15      Boyle  2009   male  physics  
16      Smith  2009   male  physics  
17      Nambu  2008   male  physics  
18  Kobayashi  2008   male  physics  
19    Maskawa  2008   male  physics

### Selecting by label
To index the dataframe by label, we use the `.loc[]` attribute. Note that there are square brackets `[]` at the end of the attribute. This makes sense since we always use square brackets to index an array. The syntax is as follows, we call `.loc[row label(s), column label(s)]`. What we are actually doing is selecting by boolean conditional here. In other words, we tell pandas to return the dataframe where certain conditions evaluate to "true". Let's see an example below with the Nobel prize data.

Let's find all the rows where the gender is female. To do this, we will have to give it row criteria by checking the assertion of whether a column, gender in our case, is equal to the value "female". Pandas will test the condtional `df['gender'] == 'female'` (which evaluates to either True or False) and return the rows where it's true. 

The row criteria is `df['gender'] == 'female'` and the column criteria is "`:`" (recall "`:`" means 'all')

In [11]:
df.loc[df['gender'] == 'female', :]

firstname   id  \
122                            Maria   79   
197                            Marie    6   
203                           Malala  914   
206                            Ellen  869   
207                           Leymah  870   
208                         Tawakkol  871   
218                     Wangari Muta  783   
219                           Shirin  773   
228                             Jody  565   
238                        Rigoberta  554   
239                 Aung San Suu Kyi  553   
248                             Alva  543   
252                    Mother Teresa  540   
256                            Betty  535   
257                          Mairead  536   
285                     Emily Greene  506   
295                             Jane  496   
323  Baroness Bertha Sophie Felicita  468   
347                           Ada E.  843   
432                 Dorothy Crowfoot  230   
469                            Irène  194   
491                            Marie    6   
502                         Svetlana  924   
504                            Alice  892   
508                            Herta  844   
510                            Doris  817   
513                         Elfriede  782   
521                          Wislawa  673   
524                             Toni  670   
526                           Nadine  668   
553                            Nelly  640   
574                         Gabriela  615   
577                            Pearl  610   
586                           Sigrid  601   
588                           Grazia  597   
604             Selma Ottilia Lovisa  579   
616                           Youyou  918   
618                        May-Britt  904   
629                     Elizabeth H.  835   
630                         Carol W.  836   
633                        Françoise  824   
643                         Linda B.  775   
663                       Christiane  453   
678                      Gertrude B.  438   
682                             Rita  435   
688                          Barbara  428   
705                          Rosalyn  413   
772                    Gerty Theresa  344   
836                           Elinor  846   

                                            motivation  share  \
122  "for their discoveries concerning nuclear shel...      4   
197  "in recognition of the extraordinary services ...      4   
203  "for their struggle against the suppression of...      2   
206  "for their non-violent struggle for the safety...      3   
207  "for their non-violent struggle for the safety...      3   
208  "for their non-violent struggle for the safety...      3   
218  "for her contribution to sustainable developme...      1   
219  "for her efforts for democracy and human right...      1   
228  "for their work for the banning and clearing o...      2   
238  "in recognition of her work for social justice...      1   
239  "for her non-violent struggle for democracy an...      1   
248                                                NaN      2   
252                                                NaN      1   
256                                                NaN      2   
257                                                NaN      2   
285                                                NaN      2   
295                                                NaN      2   
323                                                NaN      1   
347  "for studies of the structure and function of ...      3   
432  "for her determinations by X-ray techniques of...      1   
469  "in recognition of their synthesis of new radi...      2   
491  "in recognition of her services to the advance...      1   
502  "for her polyphonic writings, a monument to su...      1   
504           "master of the contemporary short story"      1   
508  "who, with the concentration of poetry and the...      1   
510  "that epicist of the female experience, who wi...      1   
513  "for her musical

If we leave the column criteria out, panda assumes we want all columns.

In [12]:
df.loc[df['gender'] == 'female'] # With column criteria left out

firstname   id  \
122                            Maria   79   
197                            Marie    6   
203                           Malala  914   
206                            Ellen  869   
207                           Leymah  870   
208                         Tawakkol  871   
218                     Wangari Muta  783   
219                           Shirin  773   
228                             Jody  565   
238                        Rigoberta  554   
239                 Aung San Suu Kyi  553   
248                             Alva  543   
252                    Mother Teresa  540   
256                            Betty  535   
257                          Mairead  536   
285                     Emily Greene  506   
295                             Jane  496   
323  Baroness Bertha Sophie Felicita  468   
347                           Ada E.  843   
432                 Dorothy Crowfoot  230   
469                            Irène  194   
491                            Marie    6   
502                         Svetlana  924   
504                            Alice  892   
508                            Herta  844   
510                            Doris  817   
513                         Elfriede  782   
521                          Wislawa  673   
524                             Toni  670   
526                           Nadine  668   
553                            Nelly  640   
574                         Gabriela  615   
577                            Pearl  610   
586                           Sigrid  601   
588                           Grazia  597   
604             Selma Ottilia Lovisa  579   
616                           Youyou  918   
618                        May-Britt  904   
629                     Elizabeth H.  835   
630                         Carol W.  836   
633                        Françoise  824   
643                         Linda B.  775   
663                       Christiane  453   
678                      Gertrude B.  438   
682                             Rita  435   
688                          Barbara  428   
705                          Rosalyn  413   
772                    Gerty Theresa  344   
836                           Elinor  846   

                                            motivation  share  \
122  "for their discoveries concerning nuclear shel...      4   
197  "in recognition of the extraordinary services ...      4   
203  "for their struggle against the suppression of...      2   
206  "for their non-violent struggle for the safety...      3   
207  "for their non-violent struggle for the safety...      3   
208  "for their non-violent struggle for the safety...      3   
218  "for her contribution to sustainable developme...      1   
219  "for her efforts for democracy and human right...      1   
228  "for their work for the banning and clearing o...      2   
238  "in recognition of her work for social justice...      1   
239  "for her non-violent struggle for democracy an...      1   
248                                                NaN      2   
252                                                NaN      1   
256                                                NaN      2   
257                                                NaN      2   
285                                                NaN      2   
295                                                NaN      2   
323                                                NaN      1   
347  "for studies of the structure and function of ...      3   
432  "for her determinations by X-ray techniques of...      1   
469  "in recognition of their synthesis of new radi...      2   
491  "in recognition of her services to the advance...      1   
502  "for her polyphonic writings, a monument to su...      1   
504           "master of the contemporary short story"      1   
508  "who, with the concentration of poetry and the...      1   
510  "that epicist of the female experience, who wi...      1   
513  "for her musical

In [13]:
df.loc[df['gender'] == 'female'].shape[0] # How depressing, only 49 prizes awarded to women. N.B. in fact, it's only 48 women since Marie Curie won twice, for two different prizes!! (a remarkable feat)

49

From above, we can confirm by looking down the gender column that we have only the female entries. Now let's do the same thing, but also subset the columns we want. Let's take gender, prize and year.

*Note that since we are subsetting by gender, we wouldn't in practice actually need to select out the gender column; it is merely to illustrate to you that we have in fact subset by gender*

In [14]:
df.loc[df['gender'] == 'female', ['gender', 'year', 'prize']]

gender  year                   prize
122  female  1963                 physics
197  female  1903                 physics
203  female  2014                   peace
206  female  2011                   peace
207  female  2011                   peace
208  female  2011                   peace
218  female  2004                   peace
219  female  2003                   peace
228  female  1997                   peace
238  female  1992                   peace
239  female  1991                   peace
248  female  1982                   peace
252  female  1979                   peace
256  female  1976                   peace
257  female  1976                   peace
285  female  1946                   peace
295  female  1931                   peace
323  female  1905                   peace
347  female  2009               chemistry
432  female  1964               chemistry
469  female  1935               chemistry
491  female  1911               chemistry
502  female  2015              literature
504  female  2013              literature
508  female  2009              literature
510  female  2007              literature
513  female  2004              literature
521  female  1996              literature
524  female  1993              literature
526  female  1991              literature
553  female  1966              literature
574  female  1945              literature
577  female  1938              literature
586  female  1928              literature
588  female  1926              literature
604  female  1909              literature
616  female  2015  physiology or medicine
618  female  2014  physiology or medicine
629  female  2009  physiology or medicine
630  female  2009  physiology or medicine
633  female  2008  physiology or medicine
643  female  2004  physiology or medicine
663  female  1995  physiology or medicine
678  female  1988  physiology or medicine
682  female  1986  physiology or medicine
688  female  1983  physiology or medicine
705  female  1977  physiology or medicine
772  female  1947  physiology or medicine
836  female  2009               economics

#### Mutiple conditionals
We can also check for multiple conditionals by using boolean logic with tuples. For example, (criteria 1) & (criteria 2). We demonstrate this below with rows that evalute to true where gender is female and prize is physics. 

In [15]:
df.loc[(df['gender'] == 'female') & (df['prize'] == 'physics')]

firstname  id                                         motivation  share  \
122     Maria  79  "for their discoveries concerning nuclear shel...      4   
197     Marie   6  "in recognition of the extraordinary services ...      4   

                   surname  year  gender    prize  
122         Goeppert Mayer  1963  female  physics  
197  Curie, née Sklodowska  1903  female  physics

This should start to generate questions in your mind. How can there only be two women to be awarded the Nobel prize in physics? I shutter to think what theoretical physics would be like without the work of Emmy Noether (among so, so many others). 

#### Numerical conditional
Now let's do the same thing but with a numerical example. Let's get only the most recent years (say 1986-present). In this case, our conditional is years greater than or equal to 1986. Where this is true, pandas will return the rows. 

In [16]:
df.loc[df['year'] >= 1986, :]

firstname   id                                         motivation  \
0         Takaaki  919  "for the discovery of neutrino oscillations, w...   
1       Arthur B.  920  "for the discovery of neutrino oscillations, w...   
2           Isamu  906  "for the invention of efficient blue light-emi...   
3         Hiroshi  907  "for the invention of efficient blue light-emi...   
4           Shuji  908  "for the invention of efficient blue light-emi...   
5        François  887  "for the theoretical discovery of a mechanism ...   
6        Peter W.  888  "for the theoretical discovery of a mechanism ...   
7           Serge  876  "for ground-breaking experimental methods that...   
8        David J.  877  "for ground-breaking experimental methods that...   
9            Saul  864  "for the discovery of the accelerating expansi...   
10       Brian P.  865  "for the discovery of the accelerating expansi...   
11        Adam G.  866  "for the discovery of the accelerating expansi...   
12          Andre  849  "for groundbreaking experiments regarding the ...   
13     Konstantin  850  "for groundbreaking experiments regarding the ...   
14   Charles Kuen  838  "for groundbreaking achievements concerning th...   
15     Willard S.  839  "for the invention of an imaging semiconductor...   
16      George E.  840  "for the invention of an imaging semiconductor...   
17       Yoichiro  826  "for the discovery of the mechanism of spontan...   
18         Makoto  827  "for the discovery of the origin of the broken...   
19      Toshihide  828  "for the discovery of the origin of the broken...   
20         Albert  814     "for the discovery of Giant Magnetoresistance"   
21          Peter  815     "for the discovery of Giant Magnetoresistance"   
22        John C.  804  "for their discovery of the blackbody form and...   
23      George F.  805  "for their discovery of the blackbody form and...   
24         Roy J.  791  "for his contribution to the quantum theory of...   
25        John L.  792  "for their contributions to the development of...   
26     Theodor W.  793  "for their contributions to the development of...   
27       David J.  776  "for the discovery of asymptotic freedom in th...   
28       H. David  777  "for the discovery of asymptotic freedom in th...   
29          Frank  778  "for the discovery of asymptotic freedom in th...   
..            ...  ...                                                ...   
847     Robert F.  771  "for methods of analyzing economic time series...   
848    Clive W.J.  772  "for methods of analyzing economic time series...   
849        Daniel  759  "for having integrated insights from psycholog...   
850     Vernon L.  760  "for having established laboratory experiments...   
851     George A.  744  "for their analyses of markets with asymmetric...   
852    A. Michael  745  "for their analyses of markets with asymmetric...   
853     Joseph E.  746  "for their analyses of markets with asymmetric...   
854      James J.  732  "for his development of theory and methods for...   
855     Daniel L.  733  "for his development of theory and methods for...   
856     Robert A.  720  "for his analysis of monetary and fiscal polic...   
857       Amartya  719       "for his contributions to welfare economics"   
858     Robert C.  717  "for a new method to determine the value of de...   
859      Myron S.  718  "for a new method to determine the value of de...   
860      James A.  715  "for their fundamental contributions to the ec...   
861       William  716  "for their fundamental contributions to the ec...   
862     Robert E.  714  "for having developed and applied the hypothes...   
863       John C.  711  "for their pioneering analysis of equilibria i...   
864       John F.  712  "for their pioneering analysis of equilibria i...   
865      Reinhard  713  "for their pioneering analysis of equilibria i...   
866     Robert W.  709  "for having renewed research in economic histo...   
867 

Notice that what we are returning is itself a dataframe. This means that we can index this dataframe, too. 

### Selecting by position
To index the dataframe by label, we use the .iloc[] attribute. Note that there are square brackets [] at the end of the attribute. This makes sense since we always use square brackets to index an array. The syntax is as follows, we call .loc[row position(s), column position(s)]. Let's see an example below with the Nobel prize data.
Let's select all rows and only the first few columns
The row criteria is ":" and the column criteria is ":3" (recall ":" means 'all')

In [17]:
df.iloc[:,:3]

firstname   id  \
0                Takaaki  919   
1              Arthur B.  920   
2                  Isamu  906   
3                Hiroshi  907   
4                  Shuji  908   
5               François  887   
6               Peter W.  888   
7                  Serge  876   
8               David J.  877   
9                   Saul  864   
10              Brian P.  865   
11               Adam G.  866   
12                 Andre  849   
13            Konstantin  850   
14          Charles Kuen  838   
15            Willard S.  839   
16             George E.  840   
17              Yoichiro  826   
18                Makoto  827   
19             Toshihide  828   
20                Albert  814   
21                 Peter  815   
22               John C.  804   
23             George F.  805   
24                Roy J.  791   
25               John L.  792   
26            Theodor W.  793   
27              David J.  776   
28              H. David  777   
29                 Frank  778   
..                   ...  ...   
870             Harry M.  704   
871            Merton H.  705   
872           William F.  706   
873               Trygve  703   
874              Maurice  702   
875            Robert M.  701   
876             James M.  700   
877               Franco  699   
878              Richard  698   
879               Gerard  697   
880            George J.  696   
881                James  695   
882          Lawrence R.  694   
883          Theodore W.  692   
884           Sir Arthur  693   
885           Herbert A.  691   
886               Bertil  689   
887             James E.  690   
888               Milton  688   
889  Leonid Vitaliyevich  686   
890          Tjalling C.  687   
891               Gunnar  684   
892     Friedrich August  685   
893              Wassily  683   
894              John R.  681   
895           Kenneth J.  682   
896                Simon  680   
897              Paul A.  679   
898               Ragnar  677   
899                  Jan  678   

                                            motivation  
0    "for the discovery of neutrino oscillations, w...  
1    "for the discovery of neutrino oscillations, w...  
2    "for the invention of efficient blue light-emi...  
3    "for the invention of efficient blue light-emi...  
4    "for the invention of efficient blue light-emi...  
5    "for the theoretical discovery of a mechanism ...  
6    "for the theoretical discovery of a mechanism ...  
7    "for ground-breaking experimental methods that...  
8    "for ground-breaking experimental methods that...  
9    "for the discovery of the accelerating expansi...  
10   "for the discovery of the accelerating expansi...  
11   "for the discovery of the accelerating expansi...  
12   "for groundbreaking experiments regarding the ...  
13   "for groundbreaking experiments regarding the ...  
14   "for groundbreaking achievements concerning th...  
15   "for the invention of an imaging semiconductor...  
16   "for the invention of an imaging semiconductor...  
17   "for the discovery of the mechanism of spontan...  
18   "for the discovery of the origin of the broken...  
19   "for the discovery of the origin of the broken...  
20      "for the discovery of Giant Magnetoresistance"  
21      "for the discovery of Giant Magnetoresistance"  
22   "for their discovery of the blackbody form and...  
23   "for their discovery of the blackbody form and...  
24   "for his contribution to the quantum theory of...  
25   "for their contributions to the development of...  
26   "for their contributions to the development of...  
27   "for the discovery of asymptotic freedom in th...  
28   "for the discovery of asymptotic freedom in th...  
29   "for the discovery of asymptotic freedom in th...  
..                                                 ...  
870  "for their pioneering work in the theory of fi...  
871  "for their pioneering work in the theory of fi...  
872  "for their pioneering wor